### PR TITLE
Allow registering multiple on_parameters_set_callback

### PR DESCRIPTION
**Pull request by ivanpauno**
(Edit)

Fixes https://github.com/ros2/rclcpp/issues/722.

The API looks like this:

```cpp
  OnSetParametersCallbackHandle::SharedPtr
  add_on_set_parameters_callback(OnParametersSetCallbackType callback);
```
```cpp
  void
  remove_on_set_parameters_callback(const OnSetParametersCallbackHandle * const handler);
```

----
(Original comment)

Based in discussion in https://github.com/ros2/rclcpp/issues/769 (TL;DR see [this comment](https://github.com/ros2/rclcpp/issues/769#issuecomment-505107740)), it would be nice to allow registering multiple callbacks with `set_on_parameters_set_callback`.

My idea was the following:

- Calling `set_on_parameters_set_callback` adds a new callback to a vector. It returns a lambda which emulates the behavior of calling all the previous registered callbacks (before, it was returning the previous callback and overriding it). If no callback was registered before, it returns `nullptr`.
- Calling `set_on_parameters_set_callback` with `nullptr` as argument clear all the registered callbacks.

Maybe it would be better to return the vector of all the callbacks instead of a lambda, but that would add a new signature (probably deprecating the old method).

Maybe, we want to implement this in a different way, or not implement it at all.
This is just a proposal of a possible implementation.

### CODE REVIEW

File: rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp [hunk 1]
```diff
@@ -526,7 +540,7 @@ NodeParameters::set_parameters_atomically(const std::vector<rclcpp::Parameter> &
       rcl_interfaces::msg::ParameterDescriptor(),  // Implicit declare uses default descriptor.
       staged_parameter_changes,
       parameter_overrides_,
-      nullptr,  // callback is explicitly null, so that it is called only once, when setting below.
+      {},  // callback is explicitly null, so that it is called only once, when setting below.
```
Comment (wjwwood) [inline rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp]: nitpick: comment should probably be updated (no longer null).

**Comment by clalancette** [comment]
> I think I would prefer a more complete solution which allowed callbacks to be registered and unregistered with random access (rather than having to clear them all with `nullptr` being the only way to remove one).
> 
> I don't know if this solution is worth doing in the meantime, since what I described above will definitely require API changes. Maybe others can comment.

I'd prefer to do the "correct" approach (for some value of correct), and then worrying about API changes later if we want to do a backport.

I agree with you that we should be able to add and remove callbacks at will without clearing out the whole list; it will make the whole feature much more robust.

**Comment by ivanpauno** [comment]
> I think I would prefer a more complete solution which allowed callbacks to be registered and unregistered with random access (rather than having to clear them all with `nullptr` being the only way to remove one).

How will the callbacks be identified? just by an integer index?
Also, I don't have an use case of random access of callbacks in mind. Could you add an example?

> I'd prefer to do the "correct" approach (for some value of correct), and then worrying about API changes later if we want to do a backport.

The proposal as-is now, I think it's not subject of backporting as is certainly breaking api. Maybe a better idea is to maintain the old behavior exactly, and add a new api for adding callbacks, accessing and removing them randomly.
We should call the callback registered in the old way, and then all the callbacks registered in the new way. That will not break the api.

Could be something like:

```
/* Adds a callback, it returns an integer indicating the identifier */
std::size_t
add_on_parameters_set_callback(OnParametersSetCallbackType callback) override;
```
```
/* Delete the callback with the provided identifier. Returns true if the identifier was valid.*/
bool
delete_on_parameters_set_callback(size_t identifier) override;
```

```
/* Get a copy of the callback registered with the provided identifier. Returns true if the identifier was valid.*/
bool
get_on_parameters_set_callback(size_t identifier, OnParametersSetCallbackType & callback) override;
```

```
/* Delete all the stored callbacks.*/
void
clear_on_parameters_set_callback() override;
```

The behavior of the existing `set_on_parameters_set_callback` and `register_param_change_callback` don't change. The callback registered by these methods is not stored in the same place as the others, and can't be accessed with the new methods. Probably, the new `clear_on_parameters_set_callback` should also delete the "old" callback.

**Comment by wjwwood** [comment]
> How will the callbacks be identified? just by an integer index?

I wouldn't use an index, as the index may change if you remove something from the list/vector. In other places we're using a "handle" object. It can be anything, maybe a pointer or some other kind of locally-unique identifier that is used as a key in a map. I think the "jump callbacks" use this pattern:

https://github.com/ros2/rclcpp/blob/0ccac1e3bd0f4314e11c86202c0c3ade57aed8ed/rclcpp/include/rclcpp/clock.hpp#L118

I even think the thing returned is a smart pointer and scoped, so another way to "unregister" the callback is to let the handle go out of scope.

> Also, I don't have an use case of random access of callbacks in mind. Could you add an example?

Normally, I would shy away from adding a feature just because someone might want it, but in this case I think that's appropriate, since I cannot imagine off the top of my head when this would be used, I know that if that case came up, then it might prevent them from using the feature all together. Also, since the code adding callbacks to a single node might come from various sources (e.g. the node itself adds the `use_sim_time` parameter callback and then user's add their own), removing all of the callbacks (as proposed in this pr) might have adverse side effects for one part or the other. Better to let them undo anything they've done without interfering with others unintentionally.

For the most part, I'd say we should do it just so the API has symmetry. I think we should avoid API's that let users create or add things without being able to remove them unless there's a specific reason for it beyond "it was easier not to have the remove interface".

> The proposal as-is now, I think it's not subject of backporting as is certainly breaking api. Maybe a better idea is to maintain the old behavior exactly, and add a new api for adding callbacks, accessing and removing them randomly.

That was my thoughts as well, add the new API that allows you to add multiple, have `set_parameter(s)` call both the single callback if set and all of the multi callback ones (populated from the new API) and then deprecate the API that sets only a single callback at a time. That way next version both will be valid ways to add a callback, but neither will interfere with the other.

> The behavior of the existing set_on_parameters_set_callback and register_param_change_callback don't change. The callback registered by these methods is not stored in the same place as the others, and can't be accessed with the new methods. Probably, the new clear_on_parameters_set_callback should also delete the "old" callback.

I didn't read this until I wrote my previous paragraph, but yes exactly this :)

Though I was imagining something more like the clock's jump callback's API:

```c++
struct OnSetParametersCallbackHandle {
  OnSetParametersCallback callback;
};

// stored in a std::vector<std::weak_ptr<OnSetParametersCallbackHandle>>

std::shared_ptr<OnSetParametersCallbackHandle>
add_on_set_parameters_callback(OnSetParametersCallback callback);

// same as handle.reset(); throws if not found
void
remove_on_set_parameters_callback(OnSetParametersCallbackHandle handle);
```

We can optionally still have `clear_on_parameters_set_callback()`, though I think it's dangerous because people will assume they are the only ones adding callbacks, and so clear should only remove "their" callbacks, but it would for instance also remove the one provided for `use_sim_time`. It would be easier to have scoped handles and the user could just let all of them go out of scope.

I would not have a `get_on_parameters_set_callback(handle/id)`, but maybe a `get_on_parameters_set_callbacks() -> const vector<weak_ptr<OnSetParametersCallbackHandle>> &` would be ok, though this is a case where it's sort of a feature looking for a problem in my opinion.

**Comment by ivanpauno** [comment]
I needed to rebase with master, to solve merge conflicts.

CI (up to test_rclcpp rclcpp, only fastrtps):
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/7574/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/3707/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6200/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/7400/)

File: rclcpp/include/rclcpp/node.hpp [hunk 2]
```diff
@@ -936,17 +938,69 @@ class Node : public std::enable_shared_from_this<Node>
...
+   *
+   * As an alternative, the unique pointer can be released:
+   *
+   *    `scoped_callback.reset()`
+   *
+   * Supposing that `scoped_callback` was the only owner.
+   *
+   * Calling `remove_on_set_parameters_callback` more than once with the same handler,
+   * or calling it after the shared pointer has been released is an error.
+   * Releasing the smart pointer after calling `remove_on_set_parameters_callback` is not a problem.
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/node.hpp]: Releasing != reset, see:

https://en.cppreference.com/w/cpp/memory/unique_ptr/release vs https://en.cppreference.com/w/cpp/memory/unique_ptr/reset, release will not destruct it. We should use consistent terminology here and above.

File: rclcpp/include/rclcpp/node.hpp [hunk 3]
```diff
@@ -936,17 +938,69 @@ class Node : public std::enable_shared_from_this<Node>
...
-   * if no callback was previously set.
+   * The callback functions must remain valid as long as the
+   * returned unique pointer is valid.
+   * The returned unique pointer can be promoted to a shared version.
+   *
+   * Releasing the returned unique pointer unregisters the callback.
+   * `remove_on_set_parameters_callback` can also be used.
+   *
+   * All the registered callbacks are called when a parameter is set.
+   * The order of the callbacks is not important.
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/node.hpp]: Is it not? What is the order called? What happens if one of them fails (returns a result with success set to false), do the others that have not been called still get called? If so, how do you combine the results if they disagree or more than one fails?

File: rclcpp/include/rclcpp/node.hpp [hunk 4]
```diff
@@ -936,17 +938,69 @@ class Node : public std::enable_shared_from_this<Node>
...
+   *    `scoped_callback.reset()`
+   *
+   * Supposing that `scoped_callback` was the only owner.
+   *
+   * Calling `remove_on_set_parameters_callback` more than once with the same handler,
+   * or calling it after the shared pointer has been released is an error.
+   * Releasing the smart pointer after calling `remove_on_set_parameters_callback` is not a problem.
+   *
+   * All the registered callbacks are called when a parameter is set.
+   * The order of the callbacks is not important.
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/node.hpp]: This seems redundant since you mention it in the add callback.

File: rclcpp/include/rclcpp/node_interfaces/node_parameters.hpp [hunk 5]
```diff
@@ -75,7 +76,9 @@ class ParameterMutationRecursionGuard
 };
 
 /// Implementation of the NodeParameters part of the Node API.
-class NodeParameters : public NodeParametersInterface
+class NodeParameters
+  : public NodeParametersInterface,
+  public std::enable_shared_from_this<NodeParameters>
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/node_interfaces/node_parameters.hpp]: This is not an ideal change, is it possible you don't actually need shared ownership of this? Would it be better to return a shared_ptr of the callback, and keep a weak ptr in this class?

File: rclcpp/include/rclcpp/node_interfaces/node_parameters.hpp [hunk 6]
```diff
@@ -180,7 +193,9 @@ class NodeParameters : public NodeParametersInterface
   // declare_parameter, etc).  In those cases, this will be set to false.
   bool parameter_modification_enabled_{true};
 
-  OnParametersSetCallbackType on_parameters_set_callback_ = nullptr;
+  OnParametersSetCallbackType on_parameters_set_callback_;
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/node_interfaces/node_parameters.hpp]: Why did you remove the initialization here?

File: rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp [hunk 7]
```diff
@@ -588,6 +615,8 @@ NodeParameters::set_parameters_atomically(const std::vector<rclcpp::Parameter> &
     // they are actually set on the official parameter storage
     parameters_,
     // this will get called once, with all the parameters to be set
+    on_parameters_set_callback_set_,
```
Comment (wjwwood) [inline rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp]: The comment for this one probably needs to be updated.

File: rclcpp/include/rclcpp/node.hpp (hunk 2 above)
Comment (wjwwood) [reply rclcpp/include/rclcpp/node.hpp]: I'd say, instead, "Resetting or letting the smart pointer go out of scope...".

File: rclcpp/include/rclcpp/node.hpp [hunk 8]
```diff
@@ -936,17 +938,69 @@ class Node : public std::enable_shared_from_this<Node>
...
-   * callback is returned when this method is used, or nullptr will be returned
-   * if no callback was previously set.
+   * The callback functions must remain valid as long as the
+   * returned unique pointer is valid.
+   * The returned unique pointer can be promoted to a shared version.
+   *
+   * Releasing the returned unique pointer unregisters the callback.
+   * `remove_on_set_parameters_callback` can also be used.
+   *
+   * All the registered callbacks are called when a parameter is set.
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/node.hpp]: Looking at the implementation, I don't think this is true, they're called in some order, and if one returns success as false, then you return immediately without running the rest.

File: rclcpp/include/rclcpp/node.hpp (hunk 2 above)
Comment (ivanpauno) [reply rclcpp/include/rclcpp/node.hpp]: Yes thanks, I sometimes mix those words.

File: rclcpp/include/rclcpp/node.hpp (hunk 3 above)
Comment (ivanpauno) [reply rclcpp/include/rclcpp/node.hpp]: I've updated the comment, I hope it's clearer now.

When rewriting it, I realized about the following:
Users are using callbacks for updating configurations when they are returning a successful result (e.g.: https://github.com/clalancette/velodyne/blob/34af494b6a8caad6e97378026e353ca9cfdaa97f/velodyne_pointcloud/src/conversions/convert.cc#L186-L219). In case other callback was registered, it may be executed later and reject the result, and the prior callback would have updated the configuration.

I wrote this thinking that each callback will check a disjoint group of parameters, but I don't know if that assumption is reasonable or not. If it is reasonable, I would just document how it works. If not, maybe having pre and post callbacks is a good idea.

File: rclcpp/include/rclcpp/node_interfaces/node_parameters.hpp (hunk 5 above)
Comment (ivanpauno) [reply rclcpp/include/rclcpp/node_interfaces/node_parameters.hpp]: I tried to avoid the possibility of UB:
https://github.com/ros2/rclcpp/blob/c7d01d7bf3a6f886577689877205ffb08ff27b42/rclcpp/src/rclcpp/clock.cpp#L128-L140

If not, I can't remove the pointer from the container in the deleter.
I have two possible work-arounds:
- Do what `clock` does, and suppose people will write the program in a way that the callback goes out of scope before the interface (so, catching `this` and not `shared_from_this` is ok). This usually happens, when inheriting from `rclcpp::Node`.
- Let the callback go out of scope, and do not delete it from the container in its deleter (use the default deleter). Store weak ptrs (and not raw ptrs), in the container. The callbacks are removed from the container if they are found to be `nullptr` when trying to call them (or, they aren't removed at all, just skipped).

Comment (ivanpauno) [reply rclcpp/include/rclcpp/node_interfaces/node_parameters.hpp]: I think we are always creating a shared pointer of `NodeParameters`.
So inheriting from `std::enable_shared_from_this` shouldn't be a problem (I should run full ci for confirming it).

File: rclcpp/include/rclcpp/node.hpp (hunk 3 above)
Comment (ivanpauno) [reply rclcpp/include/rclcpp/node.hpp]: Based on #769, I think it's a reasonable assumption.
But, there's maybe an use case where is not.

Comment (wjwwood) [reply rclcpp/include/rclcpp/node.hpp]: > I wrote this thinking that each callback will check a disjoint group of parameters, but I don't know if that assumption is reasonable or not.

I don't think that it is. The assumption of the callback must be that the values may still not be set due to another callback rejecting them. That's the only safe thing to do.

There is another set of callbacks "on change" which get called after parameters are added, updated, or removed. Currently it uses a topic and a subscription to generate these callbacks, but should be changed to use some other local mechanism (like directly calling a callback from within set parameter methods): https://github.com/ros2/rclcpp/issues/609

By the way, this pull request should probably somehow mention this issue https://github.com/ros2/rclcpp/issues/722 which it is addressing.

File: rclcpp/include/rclcpp/node_interfaces/node_parameters.hpp (hunk 5 above)
Comment (wjwwood) [reply rclcpp/include/rclcpp/node_interfaces/node_parameters.hpp]: I think both of the alternatives you suggested are better than having enable shared from this (personally). The first one is more dangerous for inexperienced users that make a mistake, but the second one is pretty safe, it just incurs the overhead of a shared pointer (versus a unique pointer) for each callback handle returned. Personally I'd go with the latter options. That mimics how callback groups and pub/sub/service's are handled within the node.

> I think we are always creating a shared pointer of NodeParameters.
So inheriting from std::enable_shared_from_this shouldn't be a problem (I should run full ci for confirming it).

That's true, but I feel that we're limiting our options (I'd like to eventually get the node interfaces as a non-shared pointer part of the node. Even if that might not be possible, this change takes us further from that goal, and I think it's not required.

Also it means that when the `NodeParameters` is destructed, all outstanding cleanup is done, and you don't have to worry about incomplete cleanup due to lingering callback handle objects.

File: rclcpp/include/rclcpp/node.hpp [hunk 9]
```diff
@@ -936,17 +938,68 @@ class Node : public std::enable_shared_from_this<Node>
...
+   *   or if it has been removed before.
+   */
+  RCLCPP_PUBLIC
+  void
+  remove_on_set_parameters_callback(const OnSetParametersCallbackHandle * const handler);
+
+  /// Register a callback to be called anytime a parameter is about to be changed.
+  /**
+   * With this method, only one callback can be set at a time. The callback that was previously
+   * set by this method is returned or `nullptr` if no callback was previously set.
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/node.hpp]: I think we should deprecate this method as well. So that we end up with (eventually) just one way to add these callbacks. I can't think of a good reason to have this method and the other two long term.

Comment (ivanpauno) [reply rclcpp/include/rclcpp/node.hpp]: Yeah, I agree.
I was thinking about doing it in a separate PR (or just in a separate commit).
We could backport everything else without adding a deprecation warning.

**Comment by ivanpauno** [comment]
@wjwwood I've addressed your comments.
I think that deprecating the all method is subject of a follow-up PR. So, this one can be backported independently. Also, I will have to open a bunch of PRs for deprecating it.
I think https://github.com/ros2/rclcpp/issues/609 is subject of another PR too.

**Comment by wjwwood** [comment]
Thanks for working on it and iterating!

**Comment by ivanpauno** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/7642/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/3754/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6249/)
* Windows [Build Status](https://ci.ros2.org/job/ci_windows/7487/)

**Comment by ivanpauno** [comment]
All CI failures are also appearing in the nightlies. Merging!
